REGRESSION (274876@main): [ iOS Debug ] accessibility/text-marker/text-marker-range-stale-node-crash.html is a consistent crash

cdumez opened the pull request:
#### d7a54ec9f06058d19cfa177b6bb2b92e6794dc5b
<pre>
REGRESSION (274876@main): [ iOS Debug ] accessibility/text-marker/text-marker-range-stale-node-crash.html is a consistent crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=274020">https://bugs.webkit.org/show_bug.cgi?id=274020</a>
<a href="https://rdar.apple.com/127901543">rdar://127901543</a>

Reviewed by Darin Adler.

Introduce SafeTextMarkerData as an alternative to TextMarkerData, which uses
a WeakPtr for `node` instead of a raw pointer. TextMarkerData currently has
to keep using a raw pointer because memcpy() / memcmp() are used with this
type&apos;s bytes. However, we should use SafeTextMarkerData wherever possible
and this patch starts doing some adoption to address the crash.

* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::characterOffsetForTextMarkerData):
(WebCore::AXObjectCache::accessibilityObjectForTextMarkerData):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::SafeTextMarkerData::toTextMarkerData const):
(WebCore::TextMarkerData::toSafeTextMarkerData const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker initWithTextMarker:cache:]):
(-[WebAccessibilityTextMarker initWithData:cache:]):
(-[WebAccessibilityTextMarker dataRepresentation]):
(-[WebAccessibilityTextMarker visiblePosition]):
(-[WebAccessibilityTextMarker characterOffset]):
(-[WebAccessibilityTextMarker isIgnored]):
(-[WebAccessibilityTextMarker accessibilityObject]):
(-[WebAccessibilityTextMarker description]):
(-[WebAccessibilityTextMarker textMarkerData]):

Canonical link: <a href="https://commits.webkit.org/278655@main">https://commits.webkit.org/278655@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64e165ab6d48eedcc5e7c7c39cbdb406c3a037dd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51218 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3555 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54475 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1908 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53521 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1582 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41710 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53317 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28152 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44151 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22828 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25471 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1407 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47447 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1482 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56071 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26328 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1368 "Found 1 new test failure: http/tests/site-isolation/window-open-with-name-cross-site.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49109 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27576 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44218 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28458 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7449 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/accessibility/AXTextMarker.h:
```diff
@@ -99,10 +101,43 @@ struct TextMarkerData {
 };
 
+// Safer version of TextMarkerData with a WeakPtr for Node.
+// TextMarkerData uses a raw pointer for Node because it is
+// used with memset / memcmp.
```
I'm kind of surprised that we are writing this as a statement of fact. Can it never change? Is there a path to wiping out this pattern? Maybe this can be modeled as serialization instead of as a `struct` relying on zeroing out padding bytes.

I might attack the problem from that end rather than having two versions of the `struct`, but it might not be as quick a way to fix this first bug.

cdumez replied:
Looks to be required for converting to and from AXTextMarkerRef but I'm not an accessibility expert. Maybe there is a way to get rid of the unsafe version completely. I was planning to follow-up to rely on the safe version as much as possible in WebKit but was planning to keep using the unsafe version for converting to and from AXTextMarkerRef (because I do not know how to avoid it).

cdumez committed:
REGRESSION (274876@main): [ iOS Debug ] accessibility/text-marker/text-marker-range-stale-node-crash.html is a consistent crash
https://bugs.webkit.org/show_bug.cgi?id=274020
rdar://127901543

Reviewed by Darin Adler.

Introduce SafeTextMarkerData as an alternative to TextMarkerData, which uses
a WeakPtr for `node` instead of a raw pointer. TextMarkerData currently has
to keep using a raw pointer because memcpy() / memcmp() are used with this
type's bytes. However, we should use SafeTextMarkerData wherever possible
and this patch starts doing some adoption to address the crash.

* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::characterOffsetForTextMarkerData):
(WebCore::AXObjectCache::accessibilityObjectForTextMarkerData):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::SafeTextMarkerData::toTextMarkerData const):
(WebCore::TextMarkerData::toSafeTextMarkerData const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker initWithTextMarker:cache:]):
(-[WebAccessibilityTextMarker initWithData:cache:]):
(-[WebAccessibilityTextMarker dataRepresentation]):
(-[WebAccessibilityTextMarker visiblePosition]):
(-[WebAccessibilityTextMarker characterOffset]):
(-[WebAccessibilityTextMarker isIgnored]):
(-[WebAccessibilityTextMarker accessibilityObject]):
(-[WebAccessibilityTextMarker description]):
(-[WebAccessibilityTextMarker textMarkerData]):

Canonical link: https://commits.webkit.org/278655@main
